File: src/actions_py/actions_py/count_until_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle
from my_robot_interfaces.action import CountUntil
 
 
class CountUntilClient(Node): # count_until_client
    def __init__(self):
        super().__init__("count_until_client") # count_until_client
        self.count_until_client_ = ActionClient(self,CountUntil,"count_until")
    
    def send_goal(self,target_number,period):
        # wait for the server
        self.count_until_client_.wait_for_server()
        
        # goal
        goal = CountUntil.Goal()
        goal.target_number = target_number
        goal.period = period
 
        # sending goal
        self.get_logger().info("Sending Goal")
        self.count_until_client_.send_goal_async(goal).add_done_callback(self.goal_response_callback)
        

    def goal_response_callback(self,future):
        self.goal_handle_:ClientGoalHandle = future.result()
        if self.goal_handle_.accepted:
            self.goal_handle_.get_result_async().add_done_callback(self.goal_result_callback)

    def goal_result_callback(self,future):
        result = future.result().result
        self.get_logger().info("Result: " + str(result.reached_number))


def main(args=None):
    rclpy.init(args=args)
    node = CountUntilClient() # count_until_client
    node.send_goal(6,1.0)
    rclpy.spin(node)
    rclpy.shutdown()
  
 
if __name__ == "__main__":
    main()

File: src/actions_cpp/src/count_until_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/count_until.hpp"
 
using CountUntil = my_robot_interfaces::action::CountUntil;
using CountUntilGoalHandle = rclcpp_action::ClientGoalHandle<CountUntil>;
using namespace std::placeholders;

class CountUntilClientNode : public rclcpp::Node 
{
public:
    CountUntilClientNode() : Node("count_until_client") 
    {
        count_until_client_ = rclcpp_action::create_client<CountUntil>(this,"count_until");
    }
    
    void send_goal(int target_number, double period){
        // Wait for the action server
        count_until_client_->wait_for_action_server();

        // create a goal
        auto goal = CountUntil::Goal();
        goal.target_number = target_number;
        goal.period = period;

        // options
        auto options = rclcpp_action::Client<CountUntil>::SendGoalOptions();
        options.goal_response_callback = std::bind(&CountUntilClientNode::goal_response_callback,this,_1);
        options.result_callback = std::bind(&CountUntilClientNode::goal_result_callback,this,_1);

        // send a goal
        RCLCPP_INFO(this->get_logger(),"Sending a Goal");
        count_until_client_->async_send_goal(goal,options);
    }
    
private:

    void goal_response_callback(const CountUntilGoalHandle::SharedPtr &goal_handle){
        if(!goal_handle){
            RCLCPP_INFO(this->get_logger(),"Goal got Rejected");
        }else{
            RCLCPP_INFO(this->get_logger(),"Goal got Accepted");
        }
    }


    void goal_result_callback(const CountUntilGoalHandle::WrappedResult &result)
    {   
        auto status = result.code;
        if (status == rclcpp_action::ResultCode::SUCCEEDED){
            RCLCPP_INFO(this->get_logger(),"Succeeded!");            
        }else if (status == rclcpp_action::ResultCode::ABORTED){
            RCLCPP_INFO(this->get_logger(),"Aborted!");            
        }  
        int reached_number = result.result->reached_number;
        RCLCPP_ERROR(this->get_logger(),"Result: %d",reached_number);
    }
    rclcpp_action::Client<CountUntil>::SharedPtr count_until_client_;

};
 
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CountUntilClientNode>(); 
    node->send_goal(3,1);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}